File: compiler/bench/mutrec3.c
// expected: 5000000
#include <stdio.h>
long g3(long n);
long h3(long n);
long f3(long n) {
    if (n == 0) return 0;
    return g3(n - 1) + 1;
}
long g3(long n) {
    if (n == 0) return 0;
    return h3(n - 1) + 1;
}
long h3(long n) {
    if (n == 0) return 0;
    return f3(n - 1) + 1;
}
int main() {
    long sum = 0;
    for (long i = 0; i < 50000; i++) {
        sum += f3(100);
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/mutrec3.fs
\ expected: 5000000
\ 3 mutually recursive functions, 5M total calls
\ f->g->h->f chain

: f3 ( n -- result ) recursive
  dup 0= if exit then
  1- g3 1+ ;

: g3 ( n -- result ) recursive
  dup 0= if exit then
  1- h3 1+ ;

: h3 ( n -- result ) recursive
  dup 0= if exit then
  1- f3 1+ ;

: main
  0 50000 0 do
    100 f3 +
  loop . cr ;
